lib/esp32/net/ethernet.toit

```diff
@@ -69,190 +55,141 @@ index df798c8..fef8c8a 100644
-      return result
-    unreachable
-
-  tcp_listen port/int -> tcp.ServerSocket:
-    with_ethernet_:
-      result := tcp_module.TcpServerSocket
-      result.listen "0.0.0.0" port
-      return result
-    unreachable
-
-  address -> net.IpAddress:
-    with_ethernet_:
-      return it.address
-    unreachable
-
-  is_closed -> bool:
-    return not open_
-
-  close_ -> none:
-    if not open_: return
-    open_ = false
-    open_count_--
-    if open_count_ > 0: return
-    ethernet := ethernet_
-    ethernet_ = null
-    ethernet.close
-
-  with_ethernet_ [block]:
-    if not open_: throw "interface is closed"
-    block.call ethernet_
-
-class EthernetDriver_:
-  logger_/log.Logger ::= log.default.with_name "ethernet"
-
-  resource_group_ := null
-
+class EthernetServiceProvider extends ServiceProvider implements ServiceHandler:
+  state_/NetworkState ::= NetworkState
+  create_resource_group_/Lambda
+
+  /**
+  The $mac_chip must be one of $MAC_CHIP_ESP32 or $MAC_CHIP_W5500.
+  The $phy_chip must be one of $PHY_CHIP_NONE, $PHY_CHIP_IP101 or $PHY_CHIP_LAN8720.
+  */
   constructor
       --phy_chip/int
-      --phy_addr/int=-1
+      --phy_address/int=-1
       --phy_reset/gpio.Pin?=null
       --mac_chip/int
       --mac_mdc/gpio.Pin?
       --mac_mdio/gpio.Pin?
       --mac_spi_device/spi.Device?
-      --mac_int/gpio.Pin?:
+      --mac_interrupt/gpio.Pin?:
     if mac_chip == MAC_CHIP_ESP32 or mac_chip == MAC_CHIP_OPENETH:
-      resource_group_ = ethernet_init_esp32_
-        mac_chip
-        phy_chip
-        phy_addr
-        (phy_reset ? phy_reset.num : -1)
-        (mac_mdc ? mac_mdc.num : -1)
-        (mac_mdio ? mac_mdio.num : -1)
+      create_resource_group_ = :: ethernet_init_esp32_
+          mac_chip
+          phy_chip
+          phy_address
+          phy_reset ? phy_reset.num : -1
+          mac_mdc ? mac_mdc.num : -1
+          mac_mdio ? mac_mdio.num : -1
+    else if phy_chip != PHY_CHIP_NONE:
+      throw "unexpected PHY chip selection"
     else:
-      if phy_chip != PHY_CHIP_NONE: throw "unexpected PHY chip selection"
-      resource_group_ = ethernet_init_spi_
-        mac_chip
-        (mac_spi_device as spi.Device_).device_
-        mac_int.num
+      create_resource_group_ = :: ethernet_init_spi_
+          mac_chip
+          (mac_spi_device as spi.Device_).device_
+          mac_interrupt.num
+    super "system/ethernet/esp32" --major=0 --minor=1
+        --tags=[NetworkService.TAG_ETHERNET]
+    provides EthernetService.SELECTOR --handler=this
+
+  handle index/int arguments/any --gid/int --client/int -> any:
+    if index == EthernetService.CONNECT_INDEX:
+      return connect client
+    if index == NetworkService.ADDRESS_INDEX:
+      network := (resource client arguments) as NetworkResource
+      return address network
+    unreachable
+
+  connect client/int -> List:
+    module ::= (state_.up: EthernetModule_ this create_resource_group_) as EthernetModule_
```

Comment:
Why do we store the result of the state_.up call in a local?

Reply:
Very good question. I think we need this in the wifi provider, so I just blindly copied it :) 

Dropped.